Mage.Sets/src/mage/cards/c/ChissGoriaForgeTyrant.java

```diff
@@ -0,0 +1,204 @@
+import mage.abilities.keyword.FlyingAbility;
+import mage.abilities.keyword.HasteAbility;
+import mage.cards.CardImpl;
+import mage.cards.CardSetInfo;
+import mage.cards.Cards;
+import mage.cards.CardsImpl;
+import mage.constants.*;
+import mage.filter.StaticFilters;
+import mage.game.Game;
+import mage.game.events.GameEvent;
+import mage.game.stack.Spell;
+import mage.players.Player;
+import mage.util.CardUtil;
+import mage.watchers.Watcher;
+
+/**
+ * @author Xanderhall
+ */
+public final class ChissGoriaForgeTyrant extends CardImpl {
+    
+    public ChissGoriaForgeTyrant(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.CREATURE}, "{6}{R}{R}{R}");
+        this.supertype.add(SuperType.LEGENDARY);
+        this.subtype.add(SubType.DRAGON);
+
+        this.power = new MageInt(5);
+        this.toughness = new MageInt(4);
+
+        // Affinity for artifacts
+        this.addAbility(new AffinityForArtifactsAbility());
+
+        // Flying, haste
+        this.addAbility(FlyingAbility.getInstance());
+        this.addAbility(HasteAbility.getInstance());
+
+        /*
+         * Whenever Chiss-Goria, Forge Tyrant attacks, exile the top five cards of your library. 
+         * You may cast an artifact spell from among them this turn. If you do, it has affinity for artifacts.
+         */
+
+        this.addAbility(new AttacksTriggeredAbility(new ChissGoriaForgeTyrantEffect()), new ChissGoriaForgeTyrantWatcher());
+    }
+
+    private ChissGoriaForgeTyrant(final ChissGoriaForgeTyrant card) {
+        super(card);
+    }
+
+    @Override
+    public ChissGoriaForgeTyrant copy() {
+        return new ChissGoriaForgeTyrant(this);
+    }
+}
+
+class ChissGoriaForgeTyrantEffect extends OneShotEffect {
+
+    ChissGoriaForgeTyrantEffect() {
+        super(Outcome.Benefit);
+        staticText = "exile the top five cards of your library. You may cast an artifact spell from among them this turn. If you do, it gains affinity for artifacts.";
+    }
+
+    private ChissGoriaForgeTyrantEffect(final ChissGoriaForgeTyrantEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public ChissGoriaForgeTyrantEffect copy() {
+        return new ChissGoriaForgeTyrantEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player player = game.getPlayer(source.getControllerId());
+        if (player == null) {
+            return false;
+        }
+
+        // Return if no cards exiled, effect still applied
+        Cards cards = new CardsImpl(player.getLibrary().getTopCards(game, 5));
+        if (cards.isEmpty()) {
+            return true;
+        }
+        player.moveCards(cards, Zone.EXILED, source, game);
+
+        Cards castableCards = new CardsImpl(cards.getCards(StaticFilters.FILTER_CARD_NON_LAND, game));
+        game.addEffect(new ChissGoriaForgeTyrantCanPlayEffect(castableCards, game), source);
```

Comment:
Okay, so I think you need to add another custom continuous effect here that handles adding the Affinity for Artifacts ability to those cards, checking the same watcher.

Needs to apply to the castableCards in exile so they can be considered playable even if not enough mana otherwise. That can be done with logic like `GainAbilityControlledSpellsEffect` but instead of checking a filter, check the morSet, and only need to check the relevant exile  zone.

Also needs to apply to the spells on the stack cast this way. I'm not sure quite how to check the approving object to be this AsThoughEffect... Serra Paragon is a reference, but that applies its ability only after spell cast event fired, but this needs to apply prior to that as costs are paid. So a workaround is to just check the same morSet but with a zcc offset of 1.

Reply:
This is one of the ideas I was thinking, I think my understanding of when and how to apply abilities just isn't up to snuff on how to do both of these things in an effect. 😅

Reply:
Okay I wrote the code I was thinking of and your unit test passes now. Can you take a look and make sure it works as expected in game?